lib/tests/formulae_dependents.rb

```diff
@@ -166,6 +166,23 @@ def install_dependent(dependent, testable_dependents, args:, build_from_source:
           test "brew", "link", dependent.full_name
         end
         test "brew", "install", "--only-dependencies", dependent.full_name
+
+        dependent.deps.each do |dependency|
+          dependency_f = dependency.to_formula
+
+          # Uninstall build- or test-only dependencies to avoid `brew linkage --test`
+          # mistakenly succeeding. Homebrew/homebrew-test-bot#611. Link them otherwise
+          # to avoid spurious CI failures.
+          subcommand = if (dependency.build? || dependency.test?) && dependency_f.any_version_installed?
+            ["uninstall", "--ignore-dependencies"]
+          elsif !dependency_f.keg_only? && !dependency_f.linked_keg.exist?
+            unlink_conflicts dependency_f
+            ["link"]
```

Comment:
So, this doesn't have to be in this PR but given `brew install` doesn't actually verify all dependents are linked (right,  correct me if I'm wrong?) does this not mean we can hide legitimate failures.

I think ideally we'd actually not require any dependency to be linked for the `brew linkage` to work (but that may be too involved to do for now).

Reply:
> So, this doesn't have to be in this PR but given `brew install` doesn't actually verify all dependents are linked (right, correct me if I'm wrong?) does this not mean we can hide legitimate failures.

Yes, I believe `brew install foo` doesn't require that everything that `foo` depends on is linked. I'm not sure I follow how this hides legitimate failures, though.

> I think ideally we'd actually not require any dependency to be linked for the `brew linkage` to work (but that may be too involved to do for now).

We actually don't require linked dependencies for `brew linkage` to work on macOS. In fact, barring the odd exception (e.g. GCC), `brew unlink foo` should never break anything on macOS. This is not true on Linux, because we don't use `opt`-paths when encoding `RPATH`s for non-keg-only dependencies, so doing `brew link` is only needed to fix Linux CI.

Fixing up `RPATH`s on Linux is probably not too involved, though I'm a little wary of changing things there since I'm much less familiar with Linux.

Reply:
> Yes, I believe `brew install foo` doesn't require that everything that `foo` depends on is linked. I'm not sure I follow how this hides legitimate failures, though.

I guess it depends if you consider `brew unlink $dependency; brew install foo; foo` failing with a linkage area to be a legitimate failure. I kinda do?

> We actually don't require linked dependencies for `brew linkage` to work on macOS. In fact, barring the odd exception (e.g. GCC), `brew unlink foo` should never break anything on macOS. This is not true on Linux, because we don't use `opt`-paths when encoding `RPATH`s for non-keg-only dependencies, so doing `brew link` is only needed to fix Linux CI.

Gotcha 👍🏻 I care less about Linux 🤣 

> Fixing up `RPATH`s on Linux is probably not too involved, though I'm a little wary of changing things there since I'm much less familiar with Linux.

Yeh, I'd be fine with this being mainly to handle the Linux case (although maybe worth gating the `link` to Linux only in that case?) but I think it'd be great to have a PR to do this `opt` change on Linux, too.